### PR TITLE
routing tables analysis wip

### PR DESCRIPTION
**Pull request by adisos**
issue #496

### CODE REVIEW

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+	return rt.implicitRT.getPath(src, dest)
+}
+
+// TODO: handle ECMP routing
+func (rt *routingTable) computeDisjointRouting() {
+	rt.nextHops = map[*ipblock.IPBlock]*ipblock.IPBlock{}
+	rt.droppedDestinations = ipblock.New()
+	rt.delegatedDestinations = ipblock.New()
+	// sort routes list by prefix length, then by priority
+	sort.Slice(rt.routesList, func(i, j int) bool {
```
Comment (zivnevo) [inline pkg/ibmvpc/routing_tables_analysis.go:285]: nit: [slices.SortFunc](https://pkg.go.dev/slices#SortFunc) typically runs faster than `sort.Slice`

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+	delegatedDestinations *ipblock.IPBlock // union of all ip-ranges for delegated destinations
+
+	// routesList is the list of routes that were added for this routing table
+	// it is used to compute the fields above
+	routesList []*route
+	implicitRT *systemImplicitRT
+}
+
+func (rt *routingTable) disjointRoutingStr() string {
```
Comment (zivnevo) [inline pkg/ibmvpc/routing_tables_analysis.go:221]: debug only? move to test file?

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+			case delegate:
+				rt.delegatedDestinations = rt.delegatedDestinations.Union(disjointDest)
+				log.Default().Printf("set %s as delegate\n", disjointDest.ToIPRanges())
+				return // skip next rules, move to the next disjoint dest
+			}
+		}
+	}
+}
+
+func (rt *routingTable) getPath(src vpcmodel.Node, dest *ipblock.IPBlock) vpcmodel.Path {
```
Comment (zivnevo) [inline pkg/ibmvpc/routing_tables_analysis.go:259]: It will make more sense if `getPath` appear after `computeDisjointRouting`

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+	"github.com/np-guard/models/pkg/ipblock"
+	"github.com/np-guard/vpc-network-config-analyzer/pkg/vpcmodel"
+)
+
+// RTAnalyzer analyzes routing in a certain vpc config
+type RTAnalyzer struct {
+	vpcConfig     *vpcmodel.VPCConfig
+	egressRT      []*egressRoutingTable
+	ingressRT     *ingressRoutingTable
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go]: right, changed to `[]*ingressRoutingTable` for now

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+	deliver action = iota // Routes the packet to the next hop target.
+	// You can add multiple routes with the same address prefix.
+	// The virtual router performs equal-cost, multi-path routing (ECMP) by using the different next hop IP addresses.
+
+	delegate // Routes the packet by using the system routing table.
+
+	drop // Drops the packet.
+
+	/*
+		Delegate-VPC - Delegates to the system's built-in routes, ignoring internet-bound routes. Required if the VPC uses non-RFC-1918 addresses
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go:94]: added

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+		destination: dest,
+		nextHop:     nextHop,
+		action:      action,
+		priority:    prio,
+	}
+
+	var err error
+	res.destIPBlock, err = ipblock.FromCidr(dest)
+	if err != nil {
+		log.Panicf("invalid dest CIDR: %e", err)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go]: done

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+		case ResourceTypePublicGateway:
+			res.pgwList = append(res.pgwList, router.(*PublicGateway))
+		case ResourceTypeFloatingIP:
+			res.fipList = append(res.fipList, router.(*FloatingIP))
+		}
+	}
+	return res
+}
+
+func newRoutingTable(routes []*route, vpcConfig *vpcmodel.VPCConfig, vpc *VPC) *routingTable {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go]: done

File: pkg/ibmvpc/routing_tables_analysis.go [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+		res.nextHopIPBlock, err = ipblock.FromCidrOrAddress(nextHop)
+		if err != nil {
+			log.Panicf("invalid next-hop CIDR: %e", err)
+		}
+	}
+
+	return res
+}
+
+func configFromVPCConfig(vpcConfig *vpcmodel.VPCConfig) *systemRTConfig {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go]: right, moved

File: pkg/ibmvpc/routing_tables_analysis.go (hunk 1 above)
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go]: changed

File: pkg/ibmvpc/routing_tables_analysis.go (hunk 2 above)
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go]: moved

File: pkg/ibmvpc/routing_tables_analysis.go (hunk 3 above)
Comment (adisos) [reply pkg/ibmvpc/routing_tables_analysis.go:258]: done

**Review by zivnevo — APPROVED** [review]

**Pull request merged**